File: pylib/bit.py
class bit:
    size = 0
    ary = None
    def __init__(self, size: int):
        self.size = size + 1
        self.ary = [0 for _ in range(size + 1)]
    
    def add(self, i: int, x: int):
        idx = i
        while idx < self.size:
            self.ary[idx] += x
    
    def sum(self, i: int) -> int:
        s = 0
        idx = i
        while idx < self.size:
            s += self.ary[idx]
        return s

File: cpplib/BIT.cpp
template <typename T>
struct BIT {
  int n;
  vector<T> bit[2];
  BIT(int n_) { init(n_); }
  void init(int n_) {
    n = n_ + 1;
    for (int p = 0; p < 2; p++) bit[p].assign(n, 0);
  }
  void add_sub(int p, int i, T x) {
    for (int idx = i; idx < n; idx += (idx & -idx)) {
      bit[p][idx] += x;
    }
  }
  void add(int l, int r, T x) {
    add_sub(0, l, -x * (l - 1));
    add_sub(0, r, x * (r - 1));
    add_sub(1, l, x);
    add_sub(1, r, -x);
  }
  T sum_sub(int p, int i) {
    T s(0);
    for (int idx = i; idx > 0; idx -= (idx & -idx)) {
      s += bit[p][idx];
    }
    return s;
  }
  T sum(int i) { return sum_sub(0, i) + sum_sub(1, i) * i; }
};